dali/operators/decoder/audio/generic_decoder.cc

```diff
@@ -0,0 +1,230 @@
+    curr_ += num;
+    return num;
+  }
+
+
+  size_t tell() {
+    return curr_;
+  }
+
+
+ private:
+  size_t length_, curr_;
+  const char *input_;
+};
+
+/*
+ * Functions: `GetFileLen`, `Seek`, `Read`, `Tell` are the callbacks required
+ * by `sf_open_fd` interface. Descriptions of these functions are copy-pasted
+ * from `libsnd` docs.
+ */
+
+/**
+ * The virtual file context must return the length of the virtual file in bytes.
+ */
+sf_count_t GetFileLen(void *This) {
+  return static_cast<MemoryStream *>(This)->length();
+}
+
+
+/**
+ *  The virtual file context must seek to offset using the seek mode provided by whence
+ *  which is one of
+ *   SEEK_CUR
+ *   SEEK_SET
+ *   SEEK_END
+ * The return value must contain the new offset in the file.
+ */
+sf_count_t Seek(sf_count_t offset, int whence, void *This) {
+  return static_cast<MemoryStream *>(This)->seek(offset, whence);
+}
+
+
+/**
+ * The virtual file context must copy ("read") "num" bytes into the buffer provided by dst
+ * and return the count of actually copied bytes.
+ */
+sf_count_t Read(void *dst, sf_count_t num, void *This) {
+  return static_cast<MemoryStream *>(This)->read(dst, num);
+}
+
+
+/**
+ * Return the current position of the virtual file context.
+ */
+sf_count_t Tell(void *This) {
+  return static_cast<MemoryStream *>(This)->tell();
+}
+
+}  // namespace
+
+
+template<typename SampleType>
+void GenericAudioDecoder<SampleType>::DecodeTyped(span<SampleType> output) {
+  impl_->DecodeTyped(output);
+}
+
+
+template<typename SampleType>
+AudioMetadata GenericAudioDecoder<SampleType>::OpenImpl(span<const char> encoded) {
+  return impl_->OpenImpl(encoded);
+}
+
+
+template<typename SampleType>
+void GenericAudioDecoder<SampleType>::CloseImpl() {
+  impl_->CloseImpl();
+}
+
+
+template<typename SampleType>
+GenericAudioDecoder<SampleType>::~GenericAudioDecoder() = default;
+
+
+template<typename SampleType>
+GenericAudioDecoder<SampleType>::GenericAudioDecoder() :
+        impl_(std::make_unique<Impl>()) {
+}
+
+
+template<typename SampleType>
+struct GenericAudioDecoder<SampleType>::Impl {
+  void DecodeTyped(span<SampleType> output) {
+    read_samples(snd_file_, output);
+  }
+
+
+  AudioMetadata OpenImpl(span<const char> encoded) {
+    assert(!encoded.empty());
+    assert(encoded.data());
+    AudioMetadata ret;
+    sf_info_ = {};
+    sf_info_.format = 0;
+    mem_stream_ = {encoded.data(), static_cast<int>(encoded.size())};
+    SF_VIRTUAL_IO sf_virtual_io = {
+            &GetFileLen,
+            &Seek,
+            &Read,
+            nullptr,  // No writing
+            &Tell
+    };
+    snd_file_ = sf_open_virtual(&sf_virtual_io, SFM_READ, &sf_info_, &mem_stream_);
+    DALI_ENFORCE(snd_file_, make_string("Failed to open encoded data: ", sf_strerror(snd_file_)));
+
+    ret.length = sf_info_.frames * sf_info_.channels;
+    ret.channels = sf_info_.channels;
+    ret.sample_rate = sf_info_.samplerate;
+    ret.channels_interleaved = false;
```

Comment:
are the channels always not interleaved?

Reply:
In this particular decoder yes

Reply:
You mean this decoder or format you are testing now?